app/models/mixins/ems_refresh_mixin.rb

```diff
@@ -0,0 +1,68 @@
+module EmsRefreshMixin
+  extend ActiveSupport::Concern
+
+  included do
+    supports :ems_refresh
+  end
+
+  class_methods do
+    def queue_refresh(ems_id_or_ids, ems_ref = nil)
+      refresh_internal(ems_id_or_ids, ems_ref, :queue => true)
+    end
+    alias_method :refresh_ems, :queue_refresh
+
+    def refresh(ems_id_or_ids, ems_ref = nil)
+      refresh_internal(ems_id_or_ids, ems_ref, :queue => false)
+    end
+
+    private
+
+    def refresh_internal(ems_id_or_object_ids, ems_ref, queue:)
+      targets =
+        if ems_ref.nil?
+          Array(ems_id_or_object_ids).map { |id| [base_class, id] }
+        else
+          ems_id = ems_id_or_object_ids
+          refresh_target(lookup_ems!(ems_id), ems_ref)
+        end
+
+      refresh_meth = queue ? :queue_refresh : :refresh
+      EmsRefresh.send(refresh_meth, targets)
+    end
+
+    def refresh_target(ems, ems_ref)
+      if ems.allow_targeted_refresh?
+        InventoryRefresh::Target.new(
+          :manager     => ems,
+          :association => refresh_association,
+          :manager_ref => {:ems_ref => ems_ref}
+        )
+      else
+        ems
+      end
+    end
+
+    def refresh_association
+      base_class.name.tableize.to_sym
+    end
+
+    def lookup_ems!(ems_id)
```

Comment:
I feel like this should move to the ExtManagementSystem class itself, because it has nothing class-specific in it, such that it needs to be in a mixin and defined on every object.